django/utils/choices.py

```diff
@@ -0,0 +1,51 @@
+from collections.abc import Callable, Iterable, Iterator, Mapping
+
+from django.db.models.enums import ChoicesMeta
+from django.utils.functional import Promise
+
+
+class ChoiceIterator:
+    """Base class for lazy iterators for choices."""
+
+
+class CallableChoiceIterator(ChoiceIterator):
+    """Iterator to lazily normalize choices generated by a callable."""
+
+    def __init__(self, choices_func):
+        self.choices_func = choices_func
+
+    def __iter__(self):
+        yield from normalize_field_choices(self.choices_func())
+
+
+def normalize_field_choices(value, *, depth=0):
+    """Normalizes choices values consistently for fields and widgets."""
+
+    match value:
+        case ChoicesMeta():
+            # Choices enumeration helpers already produce output in canonical form.
+            return value.choices
+        case ChoiceIterator() | Promise() | bytes() | str():
+            # Avoid expanding lazy choice iterators too soon by returning early.
+            # String-like types are iterable, so return early to avoid iterating.
+            return value
```

Comment:
So I believe I covered all of these cases in the tests?

`Promise` is super type for lazy objects, notably here because of translated "label" (or human-readable) values - the tests are using `gettext_lazy()` via `_(…)`. These are tested [here](https://github.com/django/django/pull/16943/commits/cbd9bf9ea839d432835842abdcdd6d0c944f48fb#diff-d485e2a2d7b8491e3211e00184bc104bcc6b19ff0d8162fe9b7ad5cc9cb826abR198-R200) when encountered at the top level and when we recurse to check if it is the nested case, if it's not nested, i.e. a "label" (or human-readble) string value, we'll return that directly even if translated (which is the case for pretty much the majority of the tests).

Reply:
I think they may not be tested because if you apply this change:

```diff
--- a/django/utils/choices.py
+++ b/django/utils/choices.py
@@ -25,7 +25,7 @@ def normalize_field_choices(value, *, depth=0):
         case ChoicesMeta():
             # Choices enumeration helpers already produce output in canonical form.
             return value.choices
-        case ChoiceIterator() | Promise() | bytes() | str():
+        case ChoiceIterator():
             # Avoid expanding lazy choice iterators too soon by returning early.
             # String-like types are iterable, so return early to avoid iterating.
             return value
```

no test fails. So I think I'm missing something since I can't understand why all the tests keep passing with the reduced case-guard. :confused:

Reply:
Oh, ho, of course. We can remove these as they're handled by the default case and we have the guard to avoid them being consumed in the `Iterable` case, i.e.

```python
        case Iterable() if not any(isinstance(x, (Promise, bytes, str)) for x in value):
```

That's why no tests fail. I'll add the comment there to explain what's happening.